### PR TITLE
add cards from the HMW pack

### PR DESCRIPTION
**Pull request by EvadLhorg**
additional engine changes for HMW cards, especially facedown cards

Co-authored by @HannoDiomedes

todo: update throneteki-json-data reference

### CODE REVIEW

File: server/game/cards/20-HMW/BraveCompanions.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            match: card => card === this,
+            effect: ability.effects.dynamicStrength(() => this.calculateStrength())
+        });
+                      
+        this.reaction({
+            when: {
+                onCardEntersPlay: event => event.card === this && event.playingType === 'marshal'
+            },
+            handler: () => {
+                this.game.promptForDeckSearch(this.controller, {
```
Comment (ystros) [inline server/game/cards/20-HMW/BraveCompanions.js:16]: I think it should be possible to implement this one using the new style game actions. See Benjen's Cache for an example.

File: server/game/cards/20-HMW/EliasLance.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+        this.reaction({
+            when: {
+                afterChallenge: event => event.challenge.loser === this.controller && this.parent.isAttacking()
+            },
+            handler: context => {
+                this.game.promptForSelect(context.event.challenge.winner, {
+                    activePromptTitle: 'Select a defending character',
+                    source: this,
+                    cardCondition: card => card.location === 'play area' &&
```
Comment (ystros) [inline server/game/cards/20-HMW/EliasLance.js:20]: Since the card uses the word "choose", this should be defined as a target, and use the `choosingPlayer` property to make the winner choose (this might require modifying what `choosingPlayer` supports)

File: server/game/cards/20-HMW/Jinglebell.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+class Jinglebell extends DrawCard {
+    setupCardAbilities() {
+        this.interrupt({
+            when: {
+                onCharacterKilled: event => event.card === this
+            },
+            handler: context => {
+                this.game.resolveGameAction(
+                    GameActions.search({
```
Comment (ystros) [inline server/game/cards/20-HMW/Jinglebell.js:12]: No need to do this nested construct. Since there's no cost or target, you can just do:
```js
gameAction: GameActions.search(context => ({
...
})
```

File: server/game/cards/20-HMW/NoSongSoSweet.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        this.reaction({
+            location: 'discard pile',
+            when: {
+                onCardPlayed: event => event.card.hasTrait('Song') && event.card.controller === this.controller
+            },
+            ignoreEventCosts: true,
+            cost: ability.costs.payGold(1),
+            handler: () => {
+                this.game.addMessage('{0} returns {1} back to their hand', this.controller, this);
+                this.controller.moveCard(this, 'hand');
```
Comment (ystros) [inline server/game/cards/20-HMW/NoSongSoSweet.js:27]: This should use `context.player`, and should use the returnCardToHand game action. (And doesn't seem to match the text on ThronesDB?)

File: server/game/cards/20-HMW/SeaSong.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                'onCardDiscarded:aggregate': event => (
+                    event.events.some(discardEvent => (
+                        discardEvent.source === 'reserve')) && 
+                    this.controller.canDraw()
+                )
+            },
+            limit: ability.limit.perRound(2),
+            choices: {
+                'Draw 1 card': () => {
+                    if(this.controller.canDraw()) {
```
Comment (ystros) [inline server/game/cards/20-HMW/SeaSong.js:21]: `context.player`

File: server/game/cards/20-HMW/TheDragonHasThreeHeads.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+const DrawCard = require('../../drawcard.js');
+
+class TheDragonHasThreeHeads extends DrawCard {
+    setupCardAbilities() {
+        this.action({
+            handler: () => {
+                let bonusMessage = [];
+
+                if(this.controller.cardsInPlay.find(card => card.name === 'Drogon')) {
```
Comment (ystros) [inline server/game/cards/20-HMW/TheDragonHasThreeHeads.js:9]: `context.player`

File: server/game/cards/20-HMW/ThoughAllMenDoDespiseUs.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+class ThoughAllMenDoDespiseUs extends DrawCard {
+    setupCardAbilities(ability) {
+        this.reaction({
+            max: ability.limit.perChallenge(1),
+            when: {
+                onDefendersDeclared: event => event.player !== this.controller && event.numOfDefendingCharacters === 0 && this.hasAttackingRaider()
+            },
+            handler: context => {
+                this.game.resolveGameAction(
+                    GameActions.simultaneously(
```
Comment (ystros) [inline server/game/cards/20-HMW/ThoughAllMenDoDespiseUs.js:13]: Can collapse into `gameAction: GameActions.simultaneously(context =>`

File: server/game/game.js [hunk 8]
```diff
@@ -1014,9 +1014,7 @@ class Game extends EventEmitter {
     }
 
     placeOnBottomOfDeck(card, options = { allowSave: true }) {
-        this.applyGameAction('placeOnBottomOfDeck', card, card => {
-            card.owner.moveCard(card, 'draw deck', { allowSave: options.allowSave, bottom: true });
-        });
+        card.owner.moveCardToBottomOfDeck(card, options.allowSave);
```
Comment (ystros) [inline server/game/game.js:1017]: Why this change? We should be moving away from all player / game methods like this and towards game actions.

File: server/game/cards/20-HMW/ALionStillHasClaws.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+                    winner: this.controller,
+                    challengeType: 'intrigue',
+                    by5: true
+                })
+            },
+            cost: ability.costs.payXGold(() => 0, () => 99),
+            handler: context => {
+                this.context = context;
+
+                this.game.promptWithMenu(this.controller, this, {
```
Comment (ystros) [inline server/game/cards/20-HMW/ALionStillHasClaws.js:26]: `context.player`

File: server/game/cards/20-HMW/AsHighAsHonor.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+const GameActions = require('../../GameActions');
+
+class AsHighAsHonor extends DrawCard {
+    setupCardAbilities(ability) {
+        this.action({
+            title: 'Put card into play',
+            phase: 'marshal',
+            cost: ability.costs.kneel(card => card.getType() === 'location' && card.isFaction('neutral') && card.getPrintedCost() >= 1),
+            target: {
+                cardCondition: card => card.location === 'hand' && card.controller === this.controller && card.isFaction('neutral') && this.controller.canPutIntoPlay(card)
```
Comment (ystros) [inline server/game/cards/20-HMW/AsHighAsHonor.js:11]: `context.player`

File: server/game/cards/20-HMW/BeneaththeGoldtheBitterSteel.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+                        match: context.target,
+                        effect: [
+                            ability.effects.addKeyword('Intimidate'),
+                            ability.effects.addKeyword('Insight')
+                        ]
+                    }));
+                    goldMessage = ' and insight and intimidate';
+                }
+                this.game.addMessage('{0} plays {1} to have {2} gain No attachments except Item and renown{3} until the end of the phase',
+                    this.controller, this, context.target, goldMessage);
```
Comment (ystros) [inline server/game/cards/20-HMW/BeneaththeGoldtheBitterSteel.js:35]: `context.player`

File: server/game/cards/20-HMW/BlackLorren.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+                    event.challenge.defenders.length >= 1
+                )
+            },
+            cost: ability.costs.discardGoldFromCard(1, card => card.getType() === 'character' && card.hasTrait('Raider') && card.controller === this.controller && card.isAttacking()),
+            target: {
+                cardCondition: card => card.isDefending()
+            },
+            handler: context => {
+                this.game.killCharacter(context.target);
+                this.game.addMessage('{0} discards 1 gold from {1} to kill {2}', this.controller, context.costs.discardToken, context.target);
```
Comment (ystros) [inline server/game/cards/20-HMW/BlackLorren.js:19]: `context.player`

File: server/game/cards/20-HMW/CornCornCorn.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+                }
+                if(!challengeTypes.includes('intrigue')) {
+                    buttons.push({ text: 'Intrigue', method: 'satisfyClaim', arg: 'intrigue' });
+                }
+                if(!challengeTypes.includes('power')) {
+                    buttons.push({ text: 'Power', method: 'satisfyClaim', arg: 'power' });
+                }
+                buttons.push({ text: 'Done', method: 'cancel' });
+
+                this.game.promptWithMenu(this.controller, this, {
```
Comment (ystros) [inline server/game/cards/20-HMW/CornCornCorn.js:34]: `context.player`

File: server/game/cards/20-HMW/EliasLance.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+                    onSelect: (player, card) => this.cardSelected(context, player, card)
+                });
+            }
+        });
+    }
+
+    cardSelected(context, player, card) {
+        context.event.challenge.winner.returnCardToHand(card);
+        this.game.addMessage('{0} uses {1} to have {2} return {3} to their hand',
+            this.controller, this, context.event.challenge.winner, card);
```
Comment (ystros) [inline server/game/cards/20-HMW/EliasLance.js:33]: `context.player`

File: server/game/cards/20-HMW/LordSteward.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+                afterChallenge: event => event.challenge.winner === this.controller && event.challenge.challengeType === 'intrigue' && this.isStewardParticipatingInChallenge()
+            },
+            target: {
+                cardCondition: card => card.location === 'play area' && card.getType() === 'character' && card.controller === this.game.currentChallenge.loser && card.getPrintedCost() <= this.getNumberOfStewards()
+            },
+            limit: ability.limit.perPhase(1),
+            handler: context => {
+                this.untilEndOfPhase(ability => ({
+                    match: context.target,
+                    effect: ability.effects.takeControl(this.controller)
```
Comment (ystros) [inline server/game/cards/20-HMW/LordSteward.js:24]: `context.player`

File: server/game/cards/20-HMW/NoSongSoSweet.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
+const DrawCard = require('../../drawcard.js');
+
+class NoSongSoSweet extends DrawCard {
+    setupCardAbilities(ability) {
+        this.action({
+            title: 'Add to challenge',
+            condition: () => this.game.isDuringChallenge({ challengeType: 'power' }),
+            target: {
+                cardCondition: card => card.location === 'play area' && !card.kneeled && card.controller === this.controller &&
```
Comment (ystros) [inline server/game/cards/20-HMW/NoSongSoSweet.js:9]: `context.player`

File: server/game/cards/20-HMW/OurBladesAreSharp.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+            targetController: 'current',
+            effect: ability.effects.reduceSelfCost('play', 1)
+        });
+        
+        this.action({
+            cost: ability.costs.sacrifice(card => card.getType() === 'character' && !card.isFaction('stark')),
+            target: {
+                mode: 'upTo',
+                numCards: 2,
+                cardCondition: card => card.controller === this.controller && card.hasTrait('House Bolton') && ['dead pile', 'discard pile'].includes(card.location)
```
Comment (ystros) [inline server/game/cards/20-HMW/OurBladesAreSharp.js:18]: `context.player`

File: server/game/cards/20-HMW/Sellsails.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+            cost: ability.costs.discardGold(),
+            limit: ability.limit.perPhase(1),
+            target: {
+                cardCondition: card => card.location === 'play area' && card.kneeled &&
+                                       (card.hasTrait('Smuggler') || card.hasTrait('Warship')),
+                gameAction: 'stand'
+            },
+            handler: context => {
+                context.target.controller.standCard(context.target);
+                this.game.addMessage('{0} discards 1 gold from {1} to stand {2}', this.controller, this, context.target);
```
Comment (ystros) [inline server/game/cards/20-HMW/Sellsails.js:16]: `context.player`

File: server/game/cards/20-HMW/SerRichardHorpe.js [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+            effect: ability.effects.dynamicStrength(() => this.power)
+        });
+
+        this.action({
+            title: 'Move 1 power',
+            limit: ability.limit.perRound(1),
+            targets: {
+                from: {
+                    activePromptTitle: 'Select card with power',
+                    cardCondition: card => ['active plot', 'faction', 'play area'].includes(card.location) && card.power > 0 && card.controller === this.controller,
```
Comment (ystros) [inline server/game/cards/20-HMW/SerRichardHorpe.js:17]: `context.player`

File: server/game/cards/20-HMW/SetDownOurDeeds.js [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+class SetDownOurDeeds extends DrawCard {
+    setupCardAbilities(ability) {
+        this.action({
+            title: 'Discard power to draw',
+            condition: () => this.controller.canDraw(),
+            cost: ability.costs.kneelFactionCard(),
+            target: {
+                activePromptTitle: 'Select card with power',
+                cardCondition: card => card.owner === this.controller && ['active plot', 'faction', 'play area'].includes(card.location) && card.power > 0,
```
Comment (ystros) [inline server/game/cards/20-HMW/SetDownOurDeeds.js:11]: `context.player`

File: server/game/cards/20-HMW/SetDownOurDeeds.js [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+                let nums = [];
+                let maxAmount = Math.min(context.target.power, 3);
+                for(var i = 1; i <= maxAmount; i++) {
+                    nums.push(i);
+                }
+                let buttons = nums.map(num => {
+                    return { text: num, method: 'numSelected', arg: num };
+                });
+
+                this.game.promptWithMenu(this.controller, this, {
```
Comment (ystros) [inline server/game/cards/20-HMW/SetDownOurDeeds.js:26]: `context.player`

**Review by ystros — COMMENTED** [review]
In addition to the below, there's a bunch of cards that could be switched to using `message:` instead of `this.game.addMessage()`, and others I didn't point out that could use `gameAction:` to simplify the cards.

File: server/game/cards/20-HMW/BraveCompanions.js (hunk 1 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/BraveCompanions.js]: fixed

File: server/game/cards/20-HMW/Jinglebell.js (hunk 3 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/Jinglebell.js]: fixed

File: server/game/cards/20-HMW/NoSongSoSweet.js (hunk 4 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/NoSongSoSweet.js]: fixed and i think it is in line with the thronesdb state.
quote from thronesdb:


Action: During a challenge, choose a standing character you control and have it participate in the challenge on your side.

Reaction: After you play a Song event, pay 1 gold to return No Song So Sweet from your discard pile to your hand.

Maybe your cache of thronesdb is out of date?

File: server/game/cards/20-HMW/SeaSong.js (hunk 5 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/SeaSong.js]: fixed

File: server/game/cards/20-HMW/TheDragonHasThreeHeads.js (hunk 6 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/TheDragonHasThreeHeads.js]: fixed

File: server/game/cards/20-HMW/ThoughAllMenDoDespiseUs.js (hunk 7 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/ThoughAllMenDoDespiseUs.js]: fixed

File: server/game/game.js (hunk 8 above)
Comment (EvadLhorg) [reply server/game/game.js:1017]: placeOnBottomOfDeck isn´t a "real" gameaction like returnCardToDeck, so i changed this helper method in game to reference the helper message in player, which in turn calls the real gameAction returnCardToDeck with the bottom: true option.
Using returnCardToDeck instead of placeOnBottomOfDeck enables Dupes to save against it (as the dupes trigger on onCardReturnedToDeck and not on placeOnBottomOfDeck). Furthermore, i changed PlankyTownTrader and the new card The Crone to reference cannotBeReturnedToDeck instead.
So all in all i try to remove placeOnBottomOfDeck as a gameAction and instead use returnCardToDeck gameaction.
While reviewing this, I noticed that i forgot to change ValarDohaeris´ reference to placeOnBottomOfDeck, will update that.

File: server/game/cards/20-HMW/ALionStillHasClaws.js (hunk 9 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/ALionStillHasClaws.js]: fixed

File: server/game/cards/20-HMW/AsHighAsHonor.js (hunk 10 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/AsHighAsHonor.js]: fixed

File: server/game/cards/20-HMW/BeneaththeGoldtheBitterSteel.js (hunk 11 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/BeneaththeGoldtheBitterSteel.js]: fixed

File: server/game/cards/20-HMW/BlackLorren.js (hunk 12 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/BlackLorren.js]: fixed

File: server/game/cards/20-HMW/CornCornCorn.js (hunk 13 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/CornCornCorn.js]: fixed

File: server/game/cards/20-HMW/EliasLance.js (hunk 14 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/EliasLance.js]: fixed

File: server/game/cards/20-HMW/LordSteward.js (hunk 15 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/LordSteward.js]: fixed

File: server/game/cards/20-HMW/NoSongSoSweet.js (hunk 16 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/NoSongSoSweet.js]: fixed

File: server/game/cards/20-HMW/OurBladesAreSharp.js (hunk 17 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/OurBladesAreSharp.js]: fixed

File: server/game/cards/20-HMW/Sellsails.js (hunk 18 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/Sellsails.js]: fixed

File: server/game/cards/20-HMW/SerRichardHorpe.js (hunk 19 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/SerRichardHorpe.js]: fixed

File: server/game/cards/20-HMW/SetDownOurDeeds.js (hunk 20 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/SetDownOurDeeds.js]: fixed

File: server/game/cards/20-HMW/SetDownOurDeeds.js (hunk 21 above)
Comment (EvadLhorg) [reply server/game/cards/20-HMW/SetDownOurDeeds.js]: fixed

File: server/game/effectengine.js [hunk 22]
```diff
@@ -87,7 +87,7 @@ class EffectEngine {
     }
 
     onCardMoved(event) {
-        this.unapplyAndRemove(effect => effect.duration === 'persistent' && effect.source === event.card && (effect.location === event.originalLocation || event.parentChanged && effect.location !== 'any'));
+        this.unapplyAndRemove(effect => effect.duration === 'persistent' && effect.source === event.card && event.originalLocation !== event.newLocation && (effect.location === event.originalLocation || event.parentChanged && effect.location !== 'any'));
```
[the remark this answers is not in the document]
Comment (EvadLhorg) [reply server/game/effectengine.js]: good catch. what i am trying to do is the following:
a card with a persistent effect like green-apple knight is facedown attached to a fool via A Mummers Farce in addition to another knight in play. the effect (STR increase) is not active while the character is facedown. as soon as the knight is marshalled, the onCardMoved event is fired, but instead of activating the effect, the effect is removed and no STR increase happens.
I will build a test for it... i think that will figuring this out easier in the future.

i changed the condition in effectEngine now to check for `!event.facedownChanged` and pass facedownChanged to the event in basecard.

File: server/game/game.js (hunk 8 above)
Comment (ystros) [reply server/game/game.js]: I ended up removing `placeOnBottomOfDeck` entirely, rather than take this change. It was only really used in 3 cards. It should make Valar Dohaeris more simultaneous (but it will still have the problem with attachments until I finish my other branch). See: https://github.com/throneteki/throneteki/pull/3136/commits/9f605fcf9c116323adf1ca3923e782c14b314b34
